### PR TITLE
Update c

### PR DESCRIPTION
**Pull request by amaanq**

### CODE REVIEW

**Comment by jdrouhard** [comment]
Why did the parser grow significantly with this version bump?

@amaanq  please tag me for reviews before merging anything here

File: grammar.js [hunk 1]
```diff
@@ -694,6 +732,15 @@ module.exports = grammar(C, {
...
 
+    _top_level_statement: ($, original) => choice(
+      original,
+      $.co_return_statement,
+      $.co_yield_statement,
+      $.for_range_loop,
+      $.try_statement,
+      $.throw_statement,
+    ),
+
```
Comment (jdrouhard) [inline grammar.js:743]: None of these are top level statements in c++

File: grammar.js [hunk 2]
```diff
@@ -249,6 +268,25 @@ module.exports = grammar(C, {
...
+      $.primitive_type,
+      $.template_type,
+      $.dependent_type,
+      $.placeholder_type_specifier,
+      $.decltype,
+      prec.right(choice(
+        alias($.qualified_type_identifier, $.qualified_identifier),
+        $._type_identifier,
+      )),
+    ),
```
Comment (jdrouhard) [inline grammar.js:288]: What is this for?

File: grammar.js [hunk 3]
```diff
@@ -125,7 +145,6 @@ module.exports = grammar(C, {
     type_qualifier: (_, original) => choice(
       original,
       'mutable',
-      'constexpr',
```
Comment (jdrouhard) [inline grammar.js:128]: This should be added back.

File: grammar.js [hunk 4]
```diff
@@ -84,6 +89,21 @@ module.exports = grammar(C, {
...
+      $.namespace_alias_definition,
+      $.using_declaration,
+      $.alias_declaration,
+      $.static_assert_declaration,
+      $.template_declaration,
+      $.template_instantiation,
+      alias($.constructor_or_destructor_definition, $.function_definition),
+      alias($.operator_cast_definition, $.function_definition),
+      alias($.operator_cast_declaration, $.declaration),
+    ),
```
Comment (jdrouhard) [inline grammar.js:105]: What is this for?

File: grammar.js [hunk 5]
```diff
@@ -810,7 +857,7 @@ module.exports = grammar(C, {
 
     // Expressions
 
-    _expression: ($, original) => choice(
+    _expression_not_binary: ($, original) => choice(
```
Comment (jdrouhard) [inline grammar.js:860]: What is the distinction between this one and a normal `_expression`? Bifurcating it like this balloons the parser size and can make it slower which is already a huge issue here. You needed to add a lot of conflicts which is typically a sign that things need to be consolidated or rewritten

File: grammar.js [hunk 6]
```diff
@@ -1191,7 +1237,6 @@ module.exports = grammar(C, {
     )),
 
     this: _ => 'this',
-    nullptr: _ => 'nullptr',
```
Comment (jdrouhard) [inline grammar.js:1194]: Why was this removed?

**Review by jdrouhard — COMMENTED** [review]
There are a lot of changes here I'm not sure should have been made. Please see my comments, and include me as a reviewer on future changes to this repo, thanks.

**Comment by amaanq** [comment]
> Why did the parser grow significantly with this version bump?
> 
> @amaanq please tag me for reviews before merging anything here

Sure, I just didn't want downstream too impacted by changing C quite a bit without having C++ adapted.

File: grammar.js (hunk 6 above)
Comment (amaanq) [reply grammar.js:1194]: C23 has `nullptr`

File: grammar.js (hunk 5 above)
Comment (amaanq) [reply grammar.js:860]: was an issue in C, top level binary expressions were conflicting with certain function declarations, I ported the changes here

File: grammar.js (hunk 3 above)
Comment (amaanq) [reply grammar.js:128]: It's in C23

File: grammar.js (hunk 2 above)
Comment (amaanq) [reply grammar.js:288]: I reworked how typedef worked to fix cases like `typedef unsigned uint;`, but I reworked it again to get rid of that and make it simpler, https://github.com/tree-sitter/tree-sitter-c/pull/152 for reference. We can update CPP to use git head for tree-sitter-c and port the changes

File: grammar.js (hunk 4 above)
Comment (amaanq) [reply grammar.js:105]: reworked top level items in C, this just adds certain bits missing to _block_item that _top_level_item had

Comment (jdrouhard) [reply grammar.js:105]: Is it possible to consolidate all these various lists of "top level-like items? It seems like there's quite a bit of copy/paste and is likely one of the reasons for the pretty large increase in state count and parser size this commit introduced.

File: grammar.js (hunk 1 above)
Comment (jdrouhard) [reply grammar.js:743]: @amaanq 

... this is identical to `_non_case_statement`. I wonder, did you tweak the c grammar to add essentially the exact same thing I added this for awhile back?

What is _top_level_statement specifically describing in the c grammar?

Comment (amaanq) [reply grammar.js:743]: Hmm you're right, they are the same. I changed expression_statement for top_level_statements because binary expressions caused issues with a declaration like so:

```c
TSLanguage *(*lang_parser)(void();
```

I just assumed top_level_statement was more or less the same, seeing that there were coroutine statements in top level statements in tests

Should I remove try and throw then?

File: grammar.js (hunk 4 above)
Comment (amaanq) [reply grammar.js:105]: Not exactly, because top level items and block items use different versions of expression statements to keep top level declarations less conflicting with binary expressions. 
https://github.com/tree-sitter/tree-sitter-c/pulls?q=is%3Apr+is%3Aclosed
